ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1917,6 +1921,48 @@ protected PutRequest createPutRequest() {
           buf.readableBytes(), BlobType.MetadataBlob,
           encryptedPerBlobKey != null ? encryptedPerBlobKey.duplicate() : null);
     }
+
+    /**
+     * Choose a random {@link PartitionId} for putting the metadata chunk and return it.
+     * @param partitionClass the partition class to choose partitions from.
+     * @param partitionIdsToExclude the list of {@link PartitionId}s that should be excluded from consideration.
+     * @return the chosen {@link PartitionId}
+     */
+    private PartitionId reserveMetadataPartition(String partitionClass, List<PartitionId> partitionIdsToExclude) {
+      PartitionId selected = clusterMap.getRandomFullyWritablePartition(partitionClass, partitionIdsToExclude);
+      if (selected == null) {
+        routerMetrics.numFailedPartitionReserveAttempts.inc();
+        return null;
+      }
+      if (!partitionClass.equals(selected.getPartitionClass())) {
+        logger.warn(
+            "{}: No partitions for partitionClass='{}' found, partitionClass='{}' used instead for metadata chunk. blobProperties={}",
+            loggingContext, partitionClass, selected.getPartitionClass(), passedInBlobProperties);
+        routerMetrics.unknownReservedPartitionClassCount.inc();
```

Comment:
Is that "unexpectedReservedPartition" better than "unknownReservedPartition"?

Reply:
changed.